Cancel recoveries even if all shards assigned

howardhuanghua opened the pull request:
## Issue
Sometimes we need to perform rolling restart for some static configurations to take effect, or rolling restart to upgrade whole cluster. We have tested one of the cluster that has 6 nodes, total 10TB data, 6000+ shards with 1 replica. Before restarting, we have done sync flush, each node cost 10+ mins to get cluster GREEN, all nodes rolling restart cost more than 1 hour. And for 100+ nodes, need more than one day to upgrade.

After sorting related logic, we found an issue, take 3 nodes A, B, C as an example:
Test version: 5.6.4, 6.4.3, 7.3.1.
Related settings:
"index.unassigned.node_left.delayed_timeout": 3m
"cluster.routing.allocation.node_concurrent_recoveries": 30
"indices.recovery.max_bytes_per_sec": 40mb

One node (A) restart flow:
- Restart node A. All shards on node A become unassigned.
- Before node A gets back, also before delay allocation timeout (3m), no unassigned shards gets relocated.
- After node A gets started, all unassigned shards start to recovery from node A, and node A gets throttled (30) soon.
- Then some unassigned shards start to be relocated to other nodes (B, C), as they are not throttled. Since peer recovery from remote node would copy segment files and translogs, shard around 30GB would cost 10+ mins (40mb/sec by default).

## Solution
With this PR optimization, one node restarting time in above case could reduce from 10+ mins to around 1 min. The main logic:
After restarted node gets back, if it gets throttled, do not relocate unassigned shards to other nodes before delay allocation timeout. Then in most case would not cause segment files copy from remote node.

howardhuanghua committed:
optimize rolling restart

howardhuanghua committed:
optimize rolling restart

elasticcla commented:
Hi @howardhuanghua, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in your Git [commit](https://github.com/elastic/elasticsearch/pull/46520.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

elasticmachine commented:
Pinging @elastic/es-distributed

DaveCTurner commented:
Thanks for the suggestion @howardhuanghua. I think we need to understand the situation you're describing a little more clearly. If your indices are successfully synced-flushed then it shouldn't matter if some of them start to recover onto other nodes because those recoveries should quickly be cancelled by recoveries onto the restarted node. Are you saying this is not the case? Can you share logs from such a restart with `logger.org.elasticsearch.gateway.GatewayAllocator: TRACE` so we can see why the recoveries are not being cancelled?

Also can you supply tests that support your change?

DaveCTurner commented:
I think I see an issue:

https://github.com/elastic/elasticsearch/blob/67e5ad2f532442fc1dd57cf92f9e4cd995082672/server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java#L399-L403

We only consider cancelling ongoing recoveries if there are unassigned shards. I think this might explain why Elasticsearch spends time completing a recovery despite the synced-flushed shard on the restarted node. Is it the case that most of the recoveries _are_ cancelled, but the last few run to completion?

howardhuanghua commented:
@DaveCTurner Yes, only few shards run to completion.
I opened logger.org.elasticsearch.gateway.GatewayAllocator: TRACE in our test cluster. And cannot see any useful log. But during the restarted node gets back, before delay timeout, I could see following INITIALIZING shards, 9.28.82.208 is the restarted node, 9.28.82.74 is the node that should not allocate shard.

lb_backend_server-300@1568044800000_7              2 r INITIALIZING                    9.28.82.74  1527044744023702309
cvm_device-10@1568044800000_1                      2 r INITIALIZING                    9.28.82.208 1527044744023702209
vbcgw_broute_tunnel-60@1567872000000_3             1 r INITIALIZING                    9.28.82.208 1527044744023702209
disk_iostat-10@1568044800000_1                     2 r INITIALIZING                    9.28.82.208 1527044744023702209

DaveCTurner commented:
> @DaveCTurner Yes, only few shards run to completion.

Thanks for confirming. In this case, I think it'd be better to contemplate cancelling recoveries even if there are no unallocated shards, rather than relying on the delayed allocation timeout as you propose.

howardhuanghua commented:
@DaveCTurner Thanks. I will continue to check the cancelling recoveries.

howardhuanghua commented:
Hi @DaveCTurner, we have checked sequence-number-based replica allocation. It could handle the first phase of rerouting unassigned shards. How do you think we still need to avoid allocating shards to other nodes before delay node left timeout in case of phase 1 has any issues? As the PR described in phase 2.

DaveCTurner commented:
I'm sorry I don't really understand the question. What is phase 2?

howardhuanghua commented:
Hi @DaveCTurner, in reroute method, the first phase tries to allocate unassigned shards to a node that already has a data copy:
https://github.com/elastic/elasticsearch/blob/67e5ad2f532442fc1dd57cf92f9e4cd995082672/server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java#L399-L403

And phase 2 tries to allocate unassigned shards to a node as matched as possible: 
https://github.com/elastic/elasticsearch/blob/67e5ad2f532442fc1dd57cf92f9e4cd995082672/server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java#L405
https://github.com/elastic/elasticsearch/blob/67e5ad2f532442fc1dd57cf92f9e4cd995082672/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java#L122
So I provided the above PR to avoid allocating shards to other nodes before delay node left timeout in phase 2.

DaveCTurner commented:
Sorry @howardhuanghua I am perhaps misunderstanding the issue you are trying to fix with this PR. Can you add a test case that this change fixes? I think that would make things clearer. You might find it helpful to look at `org.elasticsearch.cluster.routing.DelayedAllocationIT` since this suite tests related properties.

howardhuanghua committed:
add test case for delay allocation with no shard data copy

howardhuanghua committed:
add test case for delay allocation with no shard data copy

DaveCTurner commented on test/framework/src/main/java/org/elasticsearch/cluster/ESAllocationTestCase.java:
```diff
@@ -268,6 +271,9 @@ public void applyFailedShards(RoutingAllocation allocation, List<FailedShard> fa
         @Override
         public void allocateUnassigned(RoutingAllocation allocation) {
+            if (this.skipAllocation) {
+                return;
```
I don't understand this addition. As far as I can tell this makes this mock allocator behave quite differently from the real `GatewayAllocator` doesn't it? Your test only fails because of this difference in behaviour: if I comment this line out then your test passes without any changes to the production code. Can you provide a test case using the real allocator? I suggest adding to `DelayedAllocationIT` rather than here to ensure the test matches the production code more closely.

howardhuanghua replied:
I just want to simulate the case that if unassigned shard cannot be allocated in `GatewayAllocator`, then it also needs to be delayed in `ShardsAllocator`. Suppose we skip `GatewayAllocator` allocation, without this PR, if I remove the node that has replica shard, unassigned shard will be allocated immediately to the other node, with this PR, it will still be delayed until `delayed_timeout`. 

Considering the sequence based allocation decision could select correct node for allocation, so this PR will have side effects on phase 2 allocation that will delay shard allocation until _delayed_timeout_. For more infomation please reference https://github.com/elastic/elasticsearch/pull/46520#issuecomment-533178253.

DaveCTurner reviewed (changes requested):
Thanks @howardhuanghua, I think this test is not realistic because of the `skipAllocation` flag you've added to the `DelayedShardsMockGatewayAllocator`. I have left a more detailed comment inline.

howardhuanghua commented:
Sorry @DaveCTurner, I didn't add comments in time after commiting the test case, let me try to explain our idea clearly.

Let’s see we have a cluster with 3 nodes(A/B/C) and some indices, one of the indices named _test_ has 1 primary shard and 1 replica shard. The _test_ index shard layout is node A(shard p0), node B(shard r0), and node C. Set "index.unassigned.node_left.delayed_timeout" to 5 mins. 

The following steps create the scenario:
1. Write some data to _test_ index.
2. Do a sync flush.
3. Take node B down.
4. Do a force merge to merge multiple segments of _p0_ to 1. 
This make sure _p0_ has different sync-id with _r0_, also has different segment files.
And now _r0_ allocation will be delayed before _delayed_timeout_:
https://github.com/elastic/elasticsearch/blob/67e5ad2f532442fc1dd57cf92f9e4cd995082672/server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java#L232
5. Startup node B.
It's still before _delayed_timeout_, since both sync-id and segment files are different between _p0_ and _r0_, then `makeAllocationDecision `method would return _NOT_TAKEN_:https://github.com/elastic/elasticsearch/blob/67e5ad2f532442fc1dd57cf92f9e4cd995082672/server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java#L235 
This cause `gatewayAllocator `could not handle unassigned _r0_ (we called **phase 1** before):https://github.com/elastic/elasticsearch/blob/67e5ad2f532442fc1dd57cf92f9e4cd995082672/server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java#L402

And next step, unassigned _r0_ will be handled in `shardsAllocator `(we called **phase 2** before):
https://github.com/elastic/elasticsearch/blob/67e5ad2f532442fc1dd57cf92f9e4cd995082672/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java#L122
In this step, if node B gets throttled by other shards, _r0_ could be allocated to node C and cannot be cancelled, all the segment files need to be copied from _p0_.

Our above PR is going to stop allocating _r0_ to node C before allocation delay timeout in phase 2. 

However, if allocation decision is based on sequence number  (https://github.com/elastic/elasticsearch/issues/46318),  _r0_ would be allocated to node B in above phase 1 **if _p0_ on node A still has complete history operations**. If _r0_ cannot be allocated in phase 1, that means it **must have no** reusable data copy or complete history operations, we **should** relocate it to a new node immediately. In this case, our above PR will still wait until _delayed_timeout_, this may not be appropriate.

So all in all, we plan to only fix the ongoing recoveries cannot be cancelled issue as you mentioned in https://github.com/elastic/elasticsearch/pull/46520#issuecomment-529960293. If you think it's OK then we will provide the patch for this issue only. Please provide some advice if you have, thanks a lot.

DaveCTurner commented:
Thanks @howardhuanghua for your patient explanations. I think I now understand the issue more clearly, and I am more certain that it will be fixed by the seqno-based replica shard allocator that we're working on.

> This make sure p0 has different sync-id with r0, also has different segment files.

This is crucial to the issue you are seeing: the primary and replica must have absolutely nothing in common to hit this issue. If they share even a single segment (or a sync id) then we will hit the following code instead, and this will correctly throttle the allocation in the `ReplicaShardAllocator` rather than returning `NOT_TAKEN`:

https://github.com/elastic/elasticsearch/blob/49767fc1e9b4d9fead55aac0c47df1b41565aca4/server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java#L202-L216

Once the `ReplicaShardAllocator` takes sequence numbers into account we will execute this code if the primary and replica have any _operations_ in common, which they always do, so the shard will always be allocated back to the returning node.

> So all in all, we plan to only fix the ongoing recoveries cannot be cancelled issue as you mentioned in #46520 (comment). If you think it's OK then we will provide the patch for this issue only. Please provide some advice if you have, thanks a lot.

Yes, we'd very much appreciate a fix for that :)

howardhuanghua commented:
Hi @DaveCTurner, thanks for the response. I am glad that we are now on the same page :). I will provide the fix soon.

howardhuanghua committed:
fix initializing shard cancel logic issue

howardhuanghua committed:
fix initializing shard cancel logic issue

howardhuanghua commented:
Hi @DaveCTurner, I have updated the commit, added `hasInactiveShards` checking in both `AllocationService.reroute` and `GatewayAllocator.innerAllocatedUnassigned`. This would solve the case that only has initializing shard, and they are peer recovering to the new node rather than the existing data copy node. Please help to check, thank you.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -396,8 +396,8 @@ private void reroute(RoutingAllocation allocation) {
-        // now allocate all the unassigned to available nodes
-        if (allocation.routingNodes().unassigned().size() > 0) {
+        // now allocate all the unassigned to available nodes or cancel existing recoveries if we have a better match
+        if (allocation.routingNodes().unassigned().size() > 0 || allocation.routingNodes().hasInactiveShards()) {
```
I wonder: why not remove this condition entirely?

DaveCTurner reviewed (changes requested):
Thanks @howardhuanghua, can you also supply some tests for this change? We need at least an `ESIntegTestCase` showing that it does cancel the last batch of recoveries. I would guess you could add this to `org.elasticsearch.indices.recovery.IndexRecoveryIT`.

howardhuanghua committed:
add cancel new shard recovery and uses existing shard copy IT

howardhuanghua committed:
remove extra space

howardhuanghua commented:
Thanks @DaveCTurner, I have updated the commit.
1. Removed unassigned shard or initialization shard condition.
2. Added `testCancelNewShardRecoveryAndUsesExistingShardCopy` in `org.elasticsearch.indices.recovery.IndexRecoveryIT`. This IT simulates 3 nodes cluster, one of the replica data node down and back, the new shard recovery should be canceled and uses existing shard copy to recovery.

howardhuanghua committed:
fix long line issue

howardhuanghua commented:
Hi @DaveCTurner, would you please help to check the updated commit again? Thank you.

DaveCTurner commented on server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java:
```diff
@@ -323,6 +323,90 @@ public void testReplicaRecovery() throws Exception {
+        List<RecoveryState> nodeARecoveryStates = findRecoveriesForTargetNode(nodeA, recoveryStates);
+        assertThat(nodeARecoveryStates.size(), equalTo(1));
+        List<RecoveryState> nodeCRecoveryStates = findRecoveriesForTargetNode(nodeC, recoveryStates);
+        assertThat(nodeCRecoveryStates.size(), equalTo(1));
```
When I ran this test it failed here:

```
  2> REPRODUCE WITH: ./gradlew ':server:integTest' --tests "org.elasticsearch.indices.recovery.IndexRecoveryIT.testCancelNewShardRecoveryAndUsesExistingShardCopy {seed=[ECDF910E1F356F6D:FFC9E32BAD24745B]}" -Dtests.seed=ECDF910E1F356F6D -Dtests.security.manager=true -Dtests.jvms=4 -Dtests.locale=it -Dtests.timezone=America/Rio_Branco -Dcompiler.java=12 -Druntime.java=12
  2> java.lang.AssertionError: 
    Expected: <1>
         but: was <0>
        at __randomizedtesting.SeedInfo.seed([ECDF910E1F356F6D:FFC9E32BAD24745B]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
```

DaveCTurner commented on server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java:
```diff
@@ -323,6 +323,90 @@ public void testReplicaRecovery() throws Exception {
+        slowDownRecovery(shardSize);
+
+        logger.info("--> stop node B");
+        internalCluster().stopRandomNode(InternalTestCluster.nameFilter(nodeB));
```
It is better to use `internalCluster().restartNode()` which takes a `RestartCallback` whose `onNodeStopped` method is a good place to do things to the cluster while the node is stopped.

DaveCTurner reviewed (changes requested):
Thanks @howardhuanghua. The test you provided failed when I ran it locally (see below for details). It's normally a good idea to run tests like this repeatedly since they are not fully deterministic and might not fail every time. That said, it looks like it's doing roughly the right things and I left some ideas for small improvements.

DaveCTurner commented:
Also could you merge the latest master, because there are now some conflicts that need resolving.

howardhuanghua committed:
update test case

howardhuanghua committed:
fix compile issue

howardhuanghua commented:
Thanks @DaveCTurner, I have updated the test case based on your suggestion. During restarting replica shard node, hold peer recovery from primary shard to new node, and check peer recovery source/target on replica shard node stopped, finally make sure cluster green before releasing the held peer recovery. Please help to review again. Thanks for your help!

howardhuanghua committed:
update test case to wait for recovering

howardhuanghua commented:
Thanks @DaveCTurner, I have updated the test case based on your suggestion. During restarting replica shard node, hold peer recovery from primary shard to new node, and check peer recovery source/target on replica shard node stopped, finally make sure cluster green before releasing the held peer recovery. Please help to review again. Thanks for your help!

DaveCTurner commented on server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java:
```diff
@@ -324,6 +324,81 @@ public void testReplicaRecovery() throws Exception {
+            new InternalTestCluster.RestartCallback() {
+                @Override
+                public Settings onNodeStopped(String nodeName) throws Exception {
+                    assertBusy(() -> {
```
:grin: I was just about to note the missing wait here.

I think it'd be neater to wait for node A to send its `CLEAN_FILES` action instead of using an `assertBusy`. You can use another `CountDownLatch` for this.

DaveCTurner commented on server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java:
```diff
@@ -324,6 +324,81 @@ public void testReplicaRecovery() throws Exception {
+                        RecoveryResponse response = client().admin().indices().prepareRecoveries(INDEX_NAME).execute().actionGet();
+
+                        List<RecoveryState> recoveryStates = response.shardRecoveryStates().get(INDEX_NAME);
+                        List<RecoveryState> nodeARecoveryStates = findRecoveriesForTargetNode(nodeA, recoveryStates);
```
I think we do not need to say anything about the recoveries on node A. These assertions are true, but not particularly important for this test.

DaveCTurner commented on server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java:
```diff
@@ -324,6 +324,81 @@ public void testReplicaRecovery() throws Exception {
+                }
+            });
+
+        // wait for peer recovering from nodeA to nodeB to be finished
```
It took me some time to work out why this works - I suggest this comment explaining it:

```suggestion
        // wait for peer recovery from nodeA to nodeB which is a no-op recovery so it skips the CLEAN_FILES stage and hence is not blocked
```

DaveCTurner commented on server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java:
```diff
@@ -324,6 +324,81 @@ public void testReplicaRecovery() throws Exception {
+        // wait for peer recovering from nodeA to nodeB to be finished
+        ensureGreen();
+        allowToCompletePhase1Latch.countDown();
+        transportService.clearAllRules();
```
Could we finish by asserting that the shard copies are on nodes A and B? I know we wouldn't have been able to pass the `ensureGreen()` step above but it's a bit subtle and it'd be good to make this explicit.

DaveCTurner commented on server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java:
```diff
@@ -324,6 +324,81 @@ public void testReplicaRecovery() throws Exception {
+        final String nodeA = internalCluster().startNode();
+
+        logger.info("--> create index on node: {}", nodeA);
+        ByteSizeValue shardSize = createAndPopulateIndex(INDEX_NAME, 1, SHARD_COUNT, REPLICA_COUNT)
```
`shardSize` is unused:

```suggestion
        createAndPopulateIndex(INDEX_NAME, 1, SHARD_COUNT, REPLICA_COUNT).getShards()[0].getStats().getStore().size();
```

DaveCTurner commented on server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java:
```diff
@@ -324,6 +324,81 @@ public void testReplicaRecovery() throws Exception {
+        logger.info("--> start node B");
+        // force a shard recovery from nodeA to nodeB
+        final String nodeB = internalCluster().startNode();
+        Settings nodeBDataPathSettings = internalCluster().dataPathSettings(nodeB);
```
`nodeBDataPathSettings` is unused:

```suggestion
```

DaveCTurner commented on server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java:
```diff
@@ -324,6 +324,81 @@ public void testReplicaRecovery() throws Exception {
+
+        logger.info("--> start node C");
+        final String nodeC = internalCluster().startNode();
+        assertFalse(client().admin().cluster().prepareHealth().setWaitForNodes("3").get().isTimedOut());
```
I'd normally recommend the shorthand

```suggestion
        ensureStableCluster(3);
```

but I don't think this is necessary:

- `startNode()` calls `validateClusterFormed()`
- anyway it doesn't matter if node C takes a bit longer to join the cluster because we have to wait for its recovery to start which only happens after it's joined.

Therefore I think we can drop this:

```suggestion
```

DaveCTurner reviewed (changes requested):
Thanks @howardhuanghua I left a few more comments but this is looking very good.

howardhuanghua committed:
enhance test case

howardhuanghua commented:
Hi @DaveCTurner, appreciate your patient help! I have updated the test case, please help to check again.

DaveCTurner commented:
@elasticmachine please test this

DaveCTurner commented:
@elasticmachine test this please

DaveCTurner reviewed (approved):
LGTM thanks @howardhuanghua.
